File: content-solutions/backtracking/guard-robots.java
class Solution {
    private boolean[] cols, diag1, diag2;
    private int n;

    int place(int row) {
        if (row == n) return 1;
        int count = 0;
        for (int c = 0; c < n; c++) {
            if (cols[c] || diag1[row - c + n] || diag2[row + c]) continue;
            cols[c] = diag1[row - c + n] = diag2[row + c] = true;
            count += place(row + 1);
            cols[c] = diag1[row - c + n] = diag2[row + c] = false;
        }
        return count;
    }

    public int solution(int n) {
        this.n = n;
        cols = new boolean[n];
        diag1 = new boolean[2 * n];
        diag2 = new boolean[2 * n];
        return place(0);
    }
}

File: content-solutions/backtracking/guard-robots.js
function solution(n) {
  const cols = new Set();
  const diag1 = new Set();
  const diag2 = new Set();
  function place(row) {
    if (row === n) return 1;
    let count = 0;
    for (let c = 0; c < n; c += 1) {
      if (cols.has(c) || diag1.has(row - c) || diag2.has(row + c)) continue;
      cols.add(c);
      diag1.add(row - c);
      diag2.add(row + c);
      count += place(row + 1);
      cols.delete(c);
      diag1.delete(row - c);
      diag2.delete(row + c);
    }
    return count;
  }
  return place(0);
}
